src/fs/rock/RockRebuild.cc

```diff
@@ -557,6 +641,9 @@ Rock::Rebuild::freeBadEntry(const sfileno fileno, const char *eDescription)
 void
 Rock::Rebuild::swanSong()
 {
+    // do not declare rebuild completion if interrupted by shutdown
+    if (shutting_down && !loadedAndValidated())
+        return;
     debugs(47,3, HERE << "cache_dir #" << sd->index << " rebuild level: " <<
```

Comment:
> The build may not be necessarily 'failed' here

In my old change [request](https://github.com/measurement-factory/squid/pull/102#discussion_r489583719), "failed" can be interpreted as "did not load and validate everything it was supposed to". It does not matter _why_ the rebuild did not do/complete/finish what it was supposed to do. As you know, shutdown is just one of the ways the rebuild job may end prematurely (i.e. before loadedAndValidated() becomes true).

The code has been changed since that change request was created, but the concern is still essentially valid:

* Either we should not treat shutdown specially (because it is just one of several possibly premature termination reasons -- it is not special in any way) or
* We should not treat premature termination specially in _this_ PR (because we _need_ to treat a prematurely ended rebuild as a never-ending rebuild to accomplish _this_ PR goals).

You have [mentioned](https://github.com/measurement-factory/squid/pull/102#discussion_r491941694) that the shutdown behavior is new to this PR, and I do not disagree. However, premature termination in general is not new. The question is whether we should treat such premature termination specially in _this_ PR?

I suspect that _if_ we do not need to treat it specially to accomplish this PR goals, then we should not treat is specially (because it is out of scope and, I presume, we are not making anything worse by keeping it out of scope).

What I do not know for sure is whether there is some need to treat shutdown specifically (or premature termination in general) specially to accomplish this PR goals. In other words, will this PR make something worse or break something if we do _not_ modify swanSong()? What happens when shutdown is requested in the middle of a rebuild before and after our changes?

Reply:
> In other words, will this PR make something worse or break something if we do not modify swanSong()?

It will not break anything, I think - Squid will only print incomplete stats upon shutdown.
When the Rebuild job stops (normally or abnormally), it prints complete/incomplete statistics (old behavior). This PR added only one case when the Rebuild job may be stopped - reacting to the shutdown signal.

> What happens when shutdown is requested in the middle of a rebuild before and after our changes?

Before: the Rebuild job is not aware of that request - Squid exits without printing stats.
After: the Rebuild job is aware of the shutdown request. The job is terminated (calling swanSwong()), printing incomplete stats.

Reply:
Understood, thank you. I would not change swanSong() in this PR then. If others insist on a change, we will probably add code that handles _all_ premature job terminations the same way, but I do not see sufficient motivation to open that Pandora box.

Reply:
Ok, I undid this at f59ea0e.